File: python_src/3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        size = len(s)
        if size < 2:
            return size

        l, r = 0, 0
        cnt = 1
        table = set((s[0],))

        while True:
            r += 1
            if r >= size:
                break
            if s[r] not in table:
                table.add(s[r])
                tmp = r - l + 1
                cnt = cnt if cnt > tmp else tmp
            else:
                while s[r] in table:
                    table.remove(s[l])
                    l += 1
                table.add(s[r])

        return cnt
# @lc code=end

s = Solution()
print(s.lengthOfLongestSubstring("abcabcbb"))
print(s.lengthOfLongestSubstring("abcaacbb"))
print(s.lengthOfLongestSubstring("abccba"))
print(s.lengthOfLongestSubstring("bbbbb"))
print(s.lengthOfLongestSubstring("pwwkew"))

File: rust_src/3.无重复字符的最长子串.rs
/*
 * @lc app=leetcode.cn id=3 lang=rust
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
		use std::collections::HashMap;
		
		let mut set: HashMap<char, usize> = HashMap::new();
		let mut len = 0;

		for (i, c) in s.chars().enumerate()
		{
			if set.contains_key(&c)
			{
				let idx = *set.get(&c).unwrap();
				set.retain(|_k, v| *v > idx);
			}
			set.insert(c, i);
			if set.len() > len
			{
				len = set.len();
			}
		}

		return len as i32;
    }
}
// @lc code=end

